examples/test-player.rs

```diff
@@ -0,0 +1,465 @@
+                self.current_frame_sender
+                    .send(self.frame_queue[0].clone())
+                    .unwrap();
+                return true;
+            }
+        }
+
+        false
+    }
+
+    fn should_close_window(&self) -> bool {
+        unsafe { CLOSE_WINDOW }
+    }
+
+    fn render(
+        &mut self,
+        api: &RenderApi,
+        builder: &mut DisplayListBuilder,
+        resources: &mut ResourceUpdates,
+        layout_size: LayoutSize,
+        _pipeline_id: PipelineId,
+        _document_id: DocumentId,
+    ) {
+        let bounds = LayoutRect::new(LayoutPoint::zero(), layout_size);
+        let info = LayoutPrimitiveInfo::new(bounds);
+        builder.push_stacking_context(
+            &info,
+            ScrollPolicy::Scrollable,
+            None,
+            TransformStyle::Flat,
+            None,
+            MixBlendMode::Normal,
+            Vec::new(),
+        );
+
+        if self.frame_queue.len() > 0 {
+            // Assume dimensions of first frame.
+            let frame = &self.frame_queue[0];
+
+            create_or_update_planar_image(
+                api,
+                resources,
+                &mut self.y_channel_key,
+                frame.y_plane(),
+                0,
+            );
+            create_or_update_planar_image(
+                api,
+                resources,
+                &mut self.cb_channel_key,
+                frame.cb_plane(),
+                1,
+            );
+            create_or_update_planar_image(
+                api,
+                resources,
+                &mut self.cr_channel_key,
+                frame.cr_plane(),
+                2,
+            );
+
+            let aspect_ratio = frame.picture.width as f32 / frame.picture.height as f32;
+            let render_size = LayoutSize::new(
+                layout_size.width as f32,
+                layout_size.width as f32 / aspect_ratio,
+            );
+            let render_location =
+                LayoutPoint::new(0.0, (layout_size.height - render_size.height) / 2.0);
+            let info = LayoutPrimitiveInfo::with_clip_rect(
+                LayoutRect::new(render_location, render_size),
+                bounds,
+            );
+            builder.push_yuv_image(
+                &info,
+                YuvData::PlanarYCbCr(
+                    self.y_channel_key.unwrap(),
+                    self.cb_channel_key.unwrap(),
+                    self.cr_channel_key.unwrap(),
+                ),
+                YuvColorSpace::Rec601,
+                ImageRendering::Auto,
+            );
+        }
+
+        builder.pop_stacking_context();
+    }
+
+    fn on_event(
+        &mut self,
+        _event: glutin::Event,
+        _api: &RenderApi,
+        _document_id: DocumentId,
+    ) -> bool {
+        true
+    }
+
+    fn get_external_image_handler(&mut self) -> Option<Box<webrender::ExternalImageHandler>> {
+        mem::replace(&mut self.image_handler, None)
```

Comment:
Could this just be created here and returned, rather than replaced in the struct?

Reply:
Yes, if we wrap App::current_frame_sender in an Option. That's nicer, thanks!